### PR TITLE
[IMP] crm: add recurring revenues feature

### PR DESCRIPTION
**Pull request by pko-odoo**
Task ID-2283052
odoo/odoo#54194
odoo/enterprise#12423
odoo/upgrade#1463

### CODE REVIEW

File: addons/crm/data/crm_recurring_plan_data.xml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        <field name="name">Over 3 years</field>
+        <field name="number_of_months">36</field>
+    </record>
+
+    <record id="crm_recurring_plan_over_5_years" model="crm.recurring.plan">
+        <field name="name">Over 5 years </field>
+        <field name="number_of_months">60</field>
+    </record>
+
+</data></odoo>
```
Comment (std-odoo) [inline addons/crm/data/crm_recurring_plan_data.xml:24]: Add a newline here :)

File: addons/crm/views/crm_recurring_plan_views.xml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        <field name="help" type="html">
+          <p class="o_view_nocontent_smiling_face">
+            Create a Recurring Plan
+          </p><p>
+            Set Recurring Plans on Opportunities to display the contracts' renewal periodicity (e.g: Monthly, Yearly).
+          </p>
+        </field>
+    </record>
+
+</data></odoo>
```
Comment (std-odoo) [inline addons/crm/views/crm_recurring_plan_views.xml:40]: \n

File: addons/crm/models/crm_lead.py [hunk 3]
```diff
@@ -353,6 +359,16 @@ def _compute_expected_revenue(self):
         for lead in self:
             lead.expected_revenue = round((lead.planned_revenue or 0.0) * (lead.probability or 0) / 100.0, 2)
 
+    @api.depends('recurring_revenue', 'recurring_plan')
+    def _compute_expected_monthly_revenue(self):
+        for lead in self:
+            lead.expected_monthly_revenue = round((lead.recurring_revenue or 0.0) / (lead.recurring_plan.number_of_months or 1), 2)
```
Comment (std-odoo) [inline addons/crm/models/crm_lead.py:365]: Why do you need to use `round` ? (the `Monetary` field doesn't truncate the value based on the accuracy of the currency?)
(same question for `_compute_prorated_monthly_revenue`)

File: addons/crm/models/crm_recurring_plan.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+from odoo import fields, models
+
+
+class RecurringPlan(models.Model):
+    _name = "crm.recurring.plan"
+    _description = "Plan used to calculate recurring revenues"
+    _order = "sequence"
+
+    name = fields.Char('Plan Name', required=True, translate=True)
+    number_of_months = fields.Integer('# Months', required=True)
```
Comment (std-odoo) [inline addons/crm/models/crm_recurring_plan.py:13]: Maybe can you add a SQL constrains to check that this value is positive and not null ?

```python
_sql_constraints = [
    ('check_number_of_months', 'CHECK(number_of_months > 0)', 'The number of month must be strictly positive'),
]
```

File: addons/crm/models/crm_recurring_plan.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class RecurringPlan(models.Model):
+    _name = "crm.recurring.plan"
+    _description = "Plan used to calculate recurring revenues"
+    _order = "sequence"
+
+    _sql_constraints = [
+        ('check_number_of_months', 'CHECK(number_of_months > 0)', 'The number of month must be strictly positive'),
+    ]
```
Comment (dbeguin) [inline addons/crm/models/crm_recurring_plan.py:14]: sql_constraint should be placed after fields declaration

File: addons/crm/models/crm_lead.py [hunk 6]
```diff
@@ -107,8 +107,14 @@ class Lead(models.Model):
         help="Classify and analyze your lead/opportunity categories like: Training, Service")
     color = fields.Integer('Color Index', default=0)
     # Opportunity specific
-    planned_revenue = fields.Monetary('Expected Revenue', currency_field='company_currency', tracking=True)
+    planned_revenue = fields.Monetary('Expected Revenues', currency_field='company_currency', tracking=True)
     expected_revenue = fields.Monetary('Prorated Revenue', currency_field='company_currency', store=True, compute="_compute_expected_revenue")
+    expected_monthly_revenue = fields.Monetary('Expected MRR', currency_field='company_currency', store=True,
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py:112]: I would put this field after recurring_revenue and recurring_plan as it depends from both

File: addons/crm/models/crm_lead.py [hunk 7]
```diff
@@ -353,6 +359,16 @@ def _compute_expected_revenue(self):
         for lead in self:
             lead.expected_revenue = round((lead.planned_revenue or 0.0) * (lead.probability or 0) / 100.0, 2)
 
+    @api.depends('recurring_revenue', 'recurring_plan')
+    def _compute_expected_monthly_revenue(self):
+        for lead in self:
+            lead.expected_monthly_revenue = (lead.recurring_revenue or 0.0) / (lead.recurring_plan.number_of_months or 1)
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py:365]: I would add round(formula, 2)

File: addons/crm/models/crm_lead.py [hunk 8]
```diff
@@ -353,6 +359,16 @@ def _compute_expected_revenue(self):
...
+
+    @api.depends('expected_monthly_revenue', 'probability')
+    def _compute_prorated_monthly_revenue(self):
+        for lead in self:
+            lead.prorated_monthly_revenue = (lead.expected_monthly_revenue or 0.0) * (lead.probability or 0) / 100.0
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py:370]: also add round(formula, 2)

File: addons/crm/models/crm_lead.py [hunk 9]
```diff
@@ -107,8 +107,14 @@ class Lead(models.Model):
         help="Classify and analyze your lead/opportunity categories like: Training, Service")
     color = fields.Integer('Color Index', default=0)
     # Opportunity specific
-    planned_revenue = fields.Monetary('Expected Revenue', currency_field='company_currency', tracking=True)
+    planned_revenue = fields.Monetary('Expected Revenues', currency_field='company_currency', tracking=True)
     expected_revenue = fields.Monetary('Prorated Revenue', currency_field='company_currency', store=True, compute="_compute_expected_revenue")
+    expected_monthly_revenue = fields.Monetary('Expected MRR', currency_field='company_currency', store=True,
+                                               compute="_compute_expected_monthly_revenue", groups="crm.group_use_recurring_revenues")
+    prorated_monthly_revenue = fields.Monetary('Prorated MRR', currency_field='company_currency', store=True,
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py:114]: Wow, what a soup of inverted naming (not your fault here! :) 
planned_revenue = 'Expected Revenue'
expected_revenue = 'Prorated Revenue' 
"Historical Dafuck"
Anyways, this field is never used. I think we should add it in the dashboard view (same as for prorated_revenue_aggregate brol)
We'll see if we need to align to old messy naming brol or if if we take the opportunity to clean original fields naming.. or we we do nothing and let this as is.. Wait and see !

Comment (dbeguin) [reply addons/crm/models/crm_lead.py:114]: So, so we will take the opportunity to rename the old fields (watch out migration stuffs)
expected_revenue becomes prorated_revenue (the string does not change)
planned_revenue become expected_revenue ((the string does not change)
And you can keep your new fields as is :) 
cc : @tde-banana-odoo @LNAodoo

File: addons/crm/models/crm_lead.py (hunk 7 above)
Comment (pko-odoo) [reply addons/crm/models/crm_lead.py:365]: As @std-odoo pointed out it's a Monetary field, it seems it's already rounded as it should (I remember testing it)
Is there another reason to use the round method ?

Comment (dbeguin) [reply addons/crm/models/crm_lead.py:365]: indeed ! nevermind

File: addons/crm/models/crm_lead.py [hunk 10]
```diff
@@ -107,8 +107,16 @@ class Lead(models.Model):
         help="Classify and analyze your lead/opportunity categories like: Training, Service")
     color = fields.Integer('Color Index', default=0)
     # Opportunity specific
-    planned_revenue = fields.Monetary('Expected Revenue', currency_field='company_currency', tracking=True)
-    expected_revenue = fields.Monetary('Prorated Revenue', currency_field='company_currency', store=True, compute="_compute_expected_revenue")
+    expected_revenue = fields.Monetary('Expected Revenue', currency_field='company_currency', tracking=True)
+    prorated_revenue = fields.Monetary('Prorated Revenue', currency_field='company_currency', store=True, compute="_compute_prorated_revenue")
+    recurring_revenue = fields.Monetary('Recurring Revenues', currency_field='company_currency', groups="crm.group_use_recurring_revenues")
+    recurring_plan = fields.Many2one('crm.recurring.plan', string="Recurring Plan", groups="crm.group_use_recurring_revenues")
+    expected_monthly_revenue = fields.Monetary('Expected MRR', currency_field='company_currency', store=True,
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:114]: recurring_revenue_monthly ?

File: addons/crm/models/crm_lead.py [hunk 11]
```diff
@@ -107,8 +107,16 @@ class Lead(models.Model):
...
+                                               compute="_compute_expected_monthly_revenue",
+                                               groups="crm.group_use_recurring_revenues")
+    prorated_monthly_revenue = fields.Monetary('Prorated MRR', currency_field='company_currency', store=True,
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:117]: recurring_revenue_monthly_prorated ?

File: addons/crm/models/crm_lead.py [hunk 12]
```diff
@@ -348,10 +356,20 @@ def _compute_probabilities(self):
...
 
-    @api.depends('planned_revenue', 'probability')
-    def _compute_expected_revenue(self):
+    @api.depends('expected_revenue', 'probability')
+    def _compute_prorated_revenue(self):
         for lead in self:
-            lead.expected_revenue = round((lead.planned_revenue or 0.0) * (lead.probability or 0) / 100.0, 2)
+            lead.prorated_revenue = round((lead.expected_revenue or 0.0) * (lead.probability or 0) / 100.0, 2)
+
+    @api.depends('recurring_revenue', 'recurring_plan')
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:364]: Maybe use recurring_plan.number_of_months in triggers? We would like to ensure that changing plan name does not recompute all opportunities; and changing number of months does it.

File: addons/crm/models/crm_recurring_plan.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class RecurringPlan(models.Model):
+    _name = "crm.recurring.plan"
+    _description = "Plan used to calculate recurring revenues"
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_recurring_plan.py:9]: Description are generally short and module-prefixed, so maybe "CRM Recurring revenus plans" ?

File: addons/crm/security/ir.model.access.csv [hunk 14]
```diff
@@ -24,3 +24,5 @@ access_crm_lead_lost,access.crm.lead.lost,model_crm_lead_lost,sales_team.group_s
 access_crm_lead2opportunity_partner,access.crm.lead2opportunity.partner,model_crm_lead2opportunity_partner,sales_team.group_sale_salesman,1,1,1,0
 access_crm_lead2opportunity_partner_mass,access.crm.lead2opportunity.partner.mass,model_crm_lead2opportunity_partner_mass,sales_team.group_sale_salesman,1,1,1,0
 access_crm_merge_opportunity,access.crm.merge.opportunity,model_crm_merge_opportunity,sales_team.group_sale_salesman,1,1,1,0
+access_crm_recurring_plan_manager,crm.recurring.plan.manager,model_crm_recurring_plan,sales_team.group_sale_manager,1,1,1,1
```
Comment (tde-banana-odoo) [inline addons/crm/security/ir.model.access.csv:27]: Side-guidelining-notes: would be better with something like crm_recurring_plan_access_manager,crm.recurring.plan.access.manager, ... (model, type (acces rule), sub-group) 

Tde, guidelines-man.

File: addons/crm/views/crm_recurring_plan_views.xml [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+            </search>
+        </field>
+    </record>
+
+    <record id="crm_recurring_plan_action" model="ir.actions.act_window">
+        <field name="name">Recurring Plans</field>
+        <field name="res_model">crm.recurring.plan</field>
+        <field name="view_mode">tree</field>
+        <field name="help" type="html">
+          <p class="o_view_nocontent_smiling_face">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_recurring_plan_views.xml:32]: HumanLinter: indentation issue + missing eof line return ?

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**